### ISSUE TITLE
Trace sample hangs on startup

### ISSUE THREAD

**Issue by meltsufin** [spring-cloud/spring-cloud-gcp#1608]
The issue seems to be related to the combination of both trace and logging starters being present in the app.

It also seems to be a problem depending on the version of the Google Cloud Java libraries.

BROKEN:
```
<google-cloud-bom.version>0.84.0-alpha</google-cloud-bom.version>
```
WORKS:
```
<google-cloud-bom.version>0.77.0-alpha</google-cloud-bom.version>
```

**Comment by meltsufin** [comment]
Another question is, why does the sample seem to work on Travis on our master build??

**Comment by dzou** [comment]
The version switch where it breaks is between `0.77.0-alpha` and `0.78.0-alpha`.

Below is side-by-side comparison of the dependency-tree differences between the two versions:

![image](https://user-images.githubusercontent.com/3209274/55106882-5dccb200-50a6-11e9-9802-13ca5981fb06.png)

**Comment by dzou** [comment]
Did some debugging with @saturnism and @ChengyuanZhao.

We discovered the point at which it is hanging is during a call made to `GrpcTransportChannel.close()`. Based on call stack, it appeared related to logging library.
![image](https://user-images.githubusercontent.com/3209274/55114789-4a771200-50b9-11e9-9ef7-cd97f159da2a.png)

We found that If we delete the `<appender-ref ref="STACKDRIVER" />` from [logback.spring in the Sample](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/main/resources/logback-spring.xml#L10), the application no longer hangs. Therefore, we hypothesized that the problem is related to logging appender.

Ray presented theory that it could be that when the logging appender is initialized, it throws an exception, and this is silently caught which is root source of hanging.

**Comment by saturnism** [comment]
After debugging a bit more, it seems like no exception was thrown. I think Spring startup creates an initial context, and then the application context. The logging system was stopped and reset in between, so stop call was called normally[1]. But our client library blocked.

Opened an issue to learn about the awaitTermination time out choice [2]

[1] https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackLoggingSystem.java#L131
[2] https://github.com/googleapis/gax-java/issues/699
[3] https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-contrib/google-cloud-logging-logback/src/main/java/com/google/cloud/logging/logback/LoggingAppender.java

**Comment by saturnism** [comment]
I tried 0.84.0-alpha and didn't see blocking on my code. Will try sample code tomorrow. However, my code does display exception during stop & reset. This is likely due to the fact that:
1. the `close` ultimately shutdown async & there seems to be pending work that needed to be done.
1. To validate this, I changed the `close` call to `shutdownNow` calls to [`GrpcLoggingRpc`](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/GrpcLoggingRpc.java#L280) and the exception went away. I don't think the `close` call closed cleanly.

Exception:
```
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@2e887a0d[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@18816f99[Wrapped task = io.grpc.internal.DelayedClientTransport$5@2c208a1b]] rejected from java.util.concurrent.ScheduledThreadPoolExecutor@2341344e[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:340)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:562)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.execute(ScheduledThreadPoolExecutor.java:705)
	at io.grpc.internal.DelayedClientTransport.reprocess(DelayedClientTransport.java:301)
	at io.grpc.internal.ManagedChannelImpl.updateSubchannelPicker(ManagedChannelImpl.java:753)
	at io.grpc.internal.ManagedChannelImpl.access$5100(ManagedChannelImpl.java:89)
	at io.grpc.internal.ManagedChannelImpl$LbHelperImpl$1UpdateBalancingState.run(ManagedChannelImpl.java:1117)
	at io.grpc.internal.ChannelExecutor.drain(ChannelExecutor.java:73)
	at io.grpc.internal.InternalSubchannel$TransportListener.transportReady(InternalSubchannel.java:567)
	at io.grpc.netty.shaded.io.grpc.netty.ClientTransportLifecycleManager.notifyReady(ClientTransportLifecycleManager.java:43)
	at io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler$FrameListener.onSettingsRead(NettyClientHandler.java:745)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onSettingsRead(DefaultHttp2ConnectionDecoder.java:419)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$PrefaceFrameListener.onSettingsRead(DefaultHttp2ConnectionDecoder.java:637)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onSettingsRead(Http2InboundFrameLogger.java:93)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readSettingsFrame(DefaultHttp2FrameReader.java:542)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:263)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:118)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:390)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler$PrefaceDecoder.decode(Http2ConnectionHandler.java:254)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:450)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1429)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1199)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1243)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	at io.grpc.netty.shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)=
```

**Comment by meltsufin** [comment]
I wonder if this is again related to the Logback appender being re-entrant: https://github.com/googleapis/google-cloud-java/issues/3871.

**Comment by meltsufin** [comment]
Here's another comment that might be relevant to this discussion: https://github.com/spring-cloud/spring-cloud-gcp/issues/780#issuecomment-431598813

**Comment by meltsufin** [comment]
![image](https://user-images.githubusercontent.com/17732492/55172493-814c3700-5150-11e9-98e2-6e4574fce768.png)

Notice how the GRPC channel used for logging has a `ScheduledThreadPoolExecutor` that has no threads in it. That's why the execution gets rejected.

**Comment by dzou** [comment]
I think this is interesting: https://github.com/spring-cloud/spring-cloud-gcp/issues/780#issuecomment-400807664

Seems like the problem mentioned there and in #1096 noted that if the stackdriver appender is removed, then the problem is resolved. (Same here; if we remove the appender, the application will no longer hang). I wonder if all these problems are related.

**Comment by meltsufin** [comment]
Interestingly something like this fixes the problem:
```java
public class BetterLoggingAppender extends LoggingAppender {
	private Logging logging;

	@Override Logging getLogging() {
		if (this.logging == null) {
			synchronized(this) {
				if (this.logging == null) {
					this.logging = LoggingOptions.newBuilder().build().getService();
				}
			}
		}

		return this.logging;
	}
}
```

**Comment by dzou** [comment]
Interesting, @saturnism mentioned something about that code you posted before; like it was a synchronization anti-pattern. I wonder if there is a bug associated with that.

**Comment by meltsufin** [comment]
I think I got to the bottom of it. The issue is that `logging.flush()` is not called in the `stop()` method.

**Comment by dzou** [comment]
Hey, just an update as to why the tests were passing. It is because the hang only lasts for 6 minutes (that is the timeout that was configured in the GrpcChannel.close() code). Once the hang times out, the application starts up normally and the trace test passes.

In figure below, you can see that the hang lasts 6 minutes consistent with the timeout that was set in the google logging library. 

![image](https://user-images.githubusercontent.com/3209274/55242728-37c52000-5214-11e9-9647-97c5d2eb683d.png)

**Comment by meltsufin** [comment]
The fix [PR](https://github.com/googleapis/google-cloud-java/pull/4784) has been merged on the client libraries side, which should be available in the next `v0.90.0` release.
